File: src/conventions/pipeline/evaluation/meaning.ts
import type { ConditionResult } from "../evidence-bundle";
import type { Call } from "../../../engine/types";
import type {
  FactConstraint,
} from "../../core/agreement-module";
import type { TeachingLabel } from "../../core/authored-text";

// MeaningId — string, colon-namespaced (e.g., "stayman:ask-major", "bridge:nt-invite")
export type MeaningId = string;

// SemanticClassId — string, cross-module equivalence class (e.g., "bridge:game-invite")
export type SemanticClassId = string;

/** Recommendation band — authored semantic priority. */
export type RecommendationBand = "must" | "should" | "may" | "avoid";

/** How the specificity value was determined. */
export type SpecificityBasis = "derived" | "asserted" | "partial";

/** Communicative constraint dimensions for specificity derivation.
 *  Each dimension represents a type of hand information a bid communicates to partner. */
export type ConstraintDimension =
  | "suitIdentity"   // which specific suit(s) the bid promises
  | "suitLength"     // min/max cards in specific suits
  | "pointRange"     // HCP or total point bounds
  | "shapeClass"     // distributional shape (balanced, two-suited, shortage, etc.)
  | "suitRelation"   // relational constraints between suits
  | "suitQuality";   // honor holdings, suit solidity (A/K/Q count, "solid suit")

/** Operator for fact-based clause evaluation. */
export type FactOperator = "gte" | "lte" | "eq" | "range" | "boolean" | "in";

/** Operator subset for evaluated meaning clauses (excludes "in" which is resolved during evaluation). */
export type MeaningClauseOperator = Exclude<FactOperator, "in">;

/** Authored ranking metadata — what convention authors write.
 *  Excludes `specificity` and `specificityBasis`, which are derived by the
 *  pipeline via `deriveSpecificity()`. */
export interface AuthoredRankingMetadata {
  /** Authored semantic priority. Replaces implicit orderKey-as-truth. */
  readonly recommendationBand: RecommendationBand;
  /** Cross-module tiebreaker. Lower = higher priority. Never decides truth.
   *  Optional — defaults to 0 when absent. Injected by `createSurface()` builder
   *  from `ModuleContext.modulePrecedence`, or derived as 0 by the pipeline. */
  readonly modulePrecedence?: number;
  /** Declaration order within the module. Deterministic last-resort tiebreaker. */
  readonly declarationOrder: number;
}

/** Resolved ranking metadata — the frozen ranking knob matrix.
 *  Produced by the pipeline from AuthoredRankingMetadata + derived specificity. */
export interface RankingMetadata extends AuthoredRankingMetadata {
  /** More specific conventions beat general ones. Higher = more specific.
   *  Derived by deriveSpecificity() — never hand-authored. */
  readonly specificity: number;
  /** How the specificity value was determined:
   *  - "derived": all clauses reference primitive or bridge-derived facts with known constraint structure
   *  - "asserted": at least one clause references an opaque module-derived boolean fact
   *  - "partial": mix of derived and asserted clauses */
  readonly specificityBasis?: SpecificityBasis;
}

/** A single clause in a meaning proposal. */
export interface MeaningClause {
  readonly factId: string;
  readonly operator: MeaningClauseOperator;
  readonly value: number | boolean | { min: number; max: number };
  readonly satisfied: boolean;
  readonly description: string;
  readonly isPublic?: boolean;
  readonly observedValue?: number | boolean | string;
}

/** Evidence for how and why a meaning was evaluated. */
export interface EvaluationEvidence {
  readonly factDependencies: readonly string[];
  readonly evaluatedConditions: readonly (ConditionResult & {
    readonly conditionId: string;
  })[];
  readonly provenance: {
    readonly moduleId: string;
    readonly roundName?: string;
    readonly nodeName: string;
    readonly origin: "meaning-pipeline";
  };
}

/** The core meaning proposal type — what a bid means, independent of encoding. */
export interface MeaningProposal {
  readonly meaningId: MeaningId;
  readonly semanticClassId?: SemanticClassId;
  readonly moduleId: string;
  readonly clauses: readonly MeaningClause[];
  readonly ranking: RankingMetadata;
  readonly evidence: EvaluationEvidence;
  readonly sourceIntent: {
    readonly type: string;
    readonly params: Readonly<Record<string, string | number | boolean>>;
  };
  readonly modulePriority?: "preferred" | "alternative";
  /** Human-readable teaching label (e.g., "Stayman 2C", "Transfer to hearts").
   *  Threaded from BidMeaning.teachingLabel through the evaluation pipeline. */
  readonly teachingLabel?: TeachingLabel;
  /** True when this bid is alertable (derived from sourceIntent.type). */
  readonly isAlertable?: boolean;
  /** ACBL annotation type: alert, announce, or educational (derived from disclosure). */
  readonly annotationType?: "alert" | "announce" | "educational";
}

/** Recommendation band priority values for comparison (lower = higher priority). */
export const BAND_PRIORITY: Record<RecommendationBand, number> = {
  must: 0,
  should: 1,
  may: 2,
  avoid: 3,
};

/**
 * Compare two RankingMetadata values using the frozen lexicographic order.
 * Returns negative if a ranks higher, positive if b ranks higher, 0 if equal.
 *
 * Frozen ranking order:
 * 1. Hard gates (handled externally, not here)
 * 2. recommendationBand (must > should > may > avoid)
 * 3. Hand-fit score (handled externally by ranker)
 * 4. specificity (higher = more specific = ranks higher)
 * 5. modulePrecedence (lower = higher priority)
 * 6. declarationOrder (lower = earlier DFS = ranks higher)
 */
export function compareRanking(a: RankingMetadata, b: RankingMetadata): number {
  // Band comparison (lower BAND_PRIORITY = higher rank)
  const bandDiff =
    BAND_PRIORITY[a.recommendationBand] - BAND_PRIORITY[b.recommendationBand];
  if (bandDiff !== 0) return bandDiff;

  // Specificity (higher = ranks higher, so negate)
  const specDiff = b.specificity - a.specificity;
  if (specDiff !== 0) return specDiff;

  // Module precedence (lower = higher priority)
  const modDiff = (a.modulePrecedence ?? 0) - (b.modulePrecedence ?? 0);
  if (modDiff !== 0) return modDiff;

  // Intra-module order (lower = earlier DFS = ranks higher)
  return a.declarationOrder - b.declarationOrder;
}

// ---------------------------------------------------------------------------
// BidMeaning types (merged from meaning-surface.ts to break circular dep)
// ---------------------------------------------------------------------------

export interface BidMeaningClause extends FactConstraint {
  /** Deterministic identifier. Optional — auto-derived from factId:operator:value
   *  by the `createSurface()` builder or pipeline fallback. */
  readonly clauseId?: string;
  /** Author-provided rationale for why this clause exists (e.g., "fit with opener").
   *  Optional — when present, appended in parentheses to the auto-derived description.
   *  The auto-derived description is always computed from factId/operator/value. */
  readonly rationale?: string;
  /** When true, this clause is included in the alert's public constraints.
   *  Primitive hand facts (hand.*) are always public regardless of this flag.
   *  Use this for bridge-derived or module facts the bundle wants to disclose. */
  readonly isPublic?: boolean;
}

export interface BidMeaning {
  readonly meaningId: MeaningId;
  readonly semanticClassId: SemanticClassId;
  /** Module that owns this surface. Optional — injected by `createSurface()` builder
   *  from `ModuleContext.moduleId`, or falls back to `"unknown"` in the pipeline. */
  readonly moduleId?: string;
  readonly encoding: {
    readonly defaultCall: Call;
    readonly alternateEncodings?: readonly {
      call: Call;
      condition?: string;
    }[];
  };
  readonly clauses: readonly BidMeaningClause[];
  readonly ranking: AuthoredRankingMetadata;
  readonly sourceIntent: {
    readonly type: string;
    readonly params: Readonly<Record<string, string | number | boolean>>;
  };
  /** How this bid's meaning is disclosed to opponents at the table.
   *  - "alert": conventional bid, opponents must be told
   *  - "announcement": partner speaks the meaning aloud (e.g., transfers)
   *  - "standard": universally known convention, not ACBL-alerted (e.g., Stayman)
   *  - "natural": natural meaning, no disclosure needed */
  readonly disclosure: "alert" | "announcement" | "natural" | "standard";
  readonly teachingLabel: TeachingLabel;
  readonly surfaceBindings?: Readonly<Record<string, string>>;
}

// ---------------------------------------------------------------------------
// Semantic class catalog (merged from semantic-classes.ts)
// ---------------------------------------------------------------------------

/**
 * Convenience catalog of platform-owned canonical semantic class IDs.
 * Modules declare their own semantic classes as plain string constants.
 * This file is a convenience catalog, not a gating registry — adding a
 * new convention MUST NOT require editing this file.
 * (Frozen Contract #14 from Agreement Module IR spec)
 */
export const BRIDGE_SEMANTIC_CLASSES = {
  NT_OPENING: "bridge:1nt-opening",
  NT_INVITE: "bridge:nt-invite",
  NT_GAME: "bridge:to-3nt",
  SUIT_OPENING_1C: "bridge:1c-opening",
  SUIT_OPENING_1D: "bridge:1d-opening",
  SUIT_OPENING_1H: "bridge:1h-opening",
  SUIT_OPENING_1S: "bridge:1s-opening",
} as const;

File: src-tauri/crates/bridge-conventions/src/types/meaning.rs
//! Bid meaning types — what a bid means, independent of encoding.
//!
//! Mirrors TS types from `conventions/pipeline/evaluation/meaning.ts`
//! and `conventions/core/agreement-module.ts` (FactConstraint).

use bridge_engine::types::Call;
use serde::{Deserialize, Serialize};

use super::authored_text::TeachingLabel;

/// Meaning ID — string, colon-namespaced.
pub type MeaningId = String;

/// Semantic class ID — cross-module equivalence class.
pub type SemanticClassId = String;

/// How a bid's meaning is disclosed to opponents at the table.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Disclosure {
    #[serde(rename = "alert")]
    Alert,
    #[serde(rename = "announcement")]
    Announcement,
    #[serde(rename = "natural")]
    Natural,
    #[serde(rename = "standard")]
    Standard,
}

/// Recommendation band — authored semantic priority.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum RecommendationBand {
    #[serde(rename = "must")]
    Must,
    #[serde(rename = "should")]
    Should,
    #[serde(rename = "may")]
    May,
    #[serde(rename = "avoid")]
    Avoid,
}

/// Communicative constraint dimensions for specificity derivation.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum ConstraintDimension {
    #[serde(rename = "suitIdentity")]
    SuitIdentity,
    #[serde(rename = "suitLength")]
    SuitLength,
    #[serde(rename = "pointRange")]
    PointRange,
    #[serde(rename = "shapeClass")]
    ShapeClass,
    #[serde(rename = "suitRelation")]
    SuitRelation,
    #[serde(rename = "suitQuality")]
    SuitQuality,
}

/// Operator for fact-based clause evaluation.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum FactOperator {
    #[serde(rename = "gte")]
    Gte,
    #[serde(rename = "lte")]
    Lte,
    #[serde(rename = "eq")]
    Eq,
    #[serde(rename = "range")]
    Range,
    #[serde(rename = "boolean")]
    Boolean,
    #[serde(rename = "in")]
    In,
}

/// Value for a fact constraint — union of possible types.
/// Uses `serde_json::Number` to preserve integer vs float distinction.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ConstraintValue {
    Number(serde_json::Number),
    Bool(bool),
    String(String),
    Range { min: serde_json::Number, max: serde_json::Number },
    List(Vec<String>),
}

impl ConstraintValue {
    /// Create a numeric constraint value from an integer.
    pub fn int(n: i64) -> Self {
        Self::Number(serde_json::Number::from(n))
    }

    /// Create a numeric constraint value from a float.
    pub fn float(n: f64) -> Self {
        Self::Number(serde_json::Number::from_f64(n).unwrap())
    }
}

/// A fact constraint (from agreement-module.ts).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FactConstraint {
    pub fact_id: String,
    pub operator: FactOperator,
    pub value: ConstraintValue,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_public: Option<bool>,
}

/// Authored ranking metadata — what convention authors write.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AuthoredRankingMetadata {
    pub recommendation_band: RecommendationBand,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub module_precedence: Option<i32>,
    pub declaration_order: i32,
}

/// A clause in a BidMeaning surface.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BidMeaningClause {
    pub fact_id: String,
    pub operator: FactOperator,
    pub value: ConstraintValue,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub clause_id: Option<String>,
    /// Auto-derived description (e.g., "8+ HCP").
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rationale: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_public: Option<bool>,
}

/// Source intent — what the bid is trying to communicate.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SourceIntent {
    #[serde(rename = "type")]
    pub intent_type: String,
    pub params: std::collections::HashMap<String, serde_json::Value>,
}

/// Alternate encoding for a bid meaning.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AlternateEncoding {
    pub call: Call,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub condition: Option<String>,
}

/// Encoding specification for a bid meaning.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BidEncoding {
    pub default_call: Call,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub alternate_encodings: Option<Vec<AlternateEncoding>>,
}

/// A bid meaning surface — the complete specification of what a bid means.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BidMeaning {
    pub meaning_id: MeaningId,
    pub semantic_class_id: SemanticClassId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub module_id: Option<String>,
    pub encoding: BidEncoding,
    pub clauses: Vec<BidMeaningClause>,
    pub ranking: AuthoredRankingMetadata,
    pub source_intent: SourceIntent,
    pub disclosure: Disclosure,
    pub teaching_label: TeachingLabel,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub surface_bindings: Option<std::collections::HashMap<String, String>>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use bridge_engine::types::BidSuit;

    #[test]
    fn disclosure_serde() {
        assert_eq!(
            serde_json::to_string(&Disclosure::Alert).unwrap(),
            "\"alert\""
        );
        assert_eq!(
            serde_json::to_string(&Disclosure::Natural).unwrap(),
            "\"natural\""
        );
    }

    #[test]
    fn fact_operator_serde() {
        assert_eq!(
            serde_json::to_string(&FactOperator::Gte).unwrap(),
            "\"gte\""
        );
        assert_eq!(
            serde_json::to_string(&FactOperator::In).unwrap(),
            "\"in\""
        );
    }

    #[test]
    fn constraint_value_number() {
        let v = ConstraintValue::int(15);
        let json = serde_json::to_string(&v).unwrap();
        assert_eq!(json, "15");
        let back: ConstraintValue = serde_json::from_str("15").unwrap();
        assert_eq!(back, v);
    }

    #[test]
    fn constraint_value_bool() {
        let v = ConstraintValue::Bool(true);
        let json = serde_json::to_string(&v).unwrap();
        assert_eq!(json, "true");
    }

    #[test]
    fn constraint_value_range() {
        let v = ConstraintValue::Range {
            min: serde_json::Number::from(8),
            max: serde_json::Number::from(10),
        };
        let json = serde_json::to_string(&v).unwrap();
        let back: ConstraintValue = serde_json::from_str(&json).unwrap();
        assert_eq!(back, v);
    }

    #[test]
    fn constraint_value_list() {
        let v = ConstraintValue::List(vec!["hearts".to_string(), "spades".to_string()]);
        let json = serde_json::to_string(&v).unwrap();
        let back: ConstraintValue = serde_json::from_str(&json).unwrap();
        assert_eq!(back, v);
    }

    #[test]
    fn fact_constraint_roundtrip() {
        let fc = FactConstraint {
            fact_id: "hand.hcp".to_string(),
            operator: FactOperator::Range,
            value: ConstraintValue::Range {
                min: serde_json::Number::from(15),
                max: serde_json::Number::from(17),
            },
            is_public: Some(true),
        };
        let json = serde_json::to_string(&fc).unwrap();
        let back: FactConstraint = serde_json::from_str(&json).unwrap();
        assert_eq!(back, fc);
    }

    #[test]
    fn bid_meaning_minimal_roundtrip() {
        let meaning = BidMeaning {
            meaning_id: "stayman:ask-major".to_string(),
            semantic_class_id: "bridge:stayman-ask".to_string(),
            module_id: Some("stayman".to_string()),
            encoding: BidEncoding {
                default_call: Call::Bid {
                    level: 2,
                    strain: BidSuit::Clubs,
                },
                alternate_encodings: None,
            },
            clauses: vec![BidMeaningClause {
                fact_id: "hand.hcp".to_string(),
                operator: FactOperator::Gte,
                value: ConstraintValue::int(8),
                clause_id: None,
                description: None,
                rationale: None,
                is_public: None,
            }],
            ranking: AuthoredRankingMetadata {
                recommendation_band: RecommendationBand::Should,
                module_precedence: None,
                declaration_order: 0,
            },
            source_intent: SourceIntent {
                intent_type: "StaymanAsk".to_string(),
                params: std::collections::HashMap::new(),
            },
            disclosure: Disclosure::Standard,
            teaching_label: TeachingLabel {
                name: super::super::authored_text::BidName::new("Stayman"),
                summary: super::super::authored_text::BidSummary::new(
                    "Asks opener for a 4-card major",
                ),
            },
            surface_bindings: None,
        };
        let json = serde_json::to_string(&meaning).unwrap();
        let back: BidMeaning = serde_json::from_str(&json).unwrap();
        assert_eq!(back, meaning);
    }
}
